PR to review master

xjuan opened the pull request:
Dummy PR to review master branch in one PR

Commit:
EosWebView: Initial implementation using GtkOffscreenWindow

Commit:
Use an ofscreen GdkWindow for each child instead of a GtkOffscreenWindow

This mean that we no longer need to marshall events from the HTMl elements
since we can catch them and let Gtk send them to the offscreen window by implementing
"pick-embedded-child" signal handler

Commit:
Use getBoundingClientRect() to calculate element positions

Commit:
EosWebView: remove size_allocate default handler.

We need to handle resize request separately

Commit:
JS utils: export _js_get_string()

Commit:
EosWebView: various improvements in drawing and size allocation

Replace draw() method with damage_event()

Use a DOM observer to initialize canvas elements an call allocate them
and trigger position updates on any DOM change.

Commit:
Improve widget visibility handling

Hide all canvases by default and sync its visibility with the widgets

Commit:
EosWebView: ignore button press/release events that are for children.

Events are not consumed by children so they bubble up to the webview
but we should ignore them.

Commit:
EosWebView: throttle position update to 10Hz

Commit:
Avoid JS error on child_draw()

Sometimes while resizing a child we get an JS error
"InvalidStateError: The object is in an invalid state."
Probably because draw_child() gets called before the last one actually
finishes

Commit:
Implement unrealize() method

Commit:
Reduce the number of JS<->Native code messages

Reworked position and allocate messages to take an array of objects

Commit:
Improve child drawing from damage events

Only draw the area that got damaged instead of the whole offscreen

Create pixbufs at damage-event handler and added pixbuf id to
eosimagedata uri handler which allows to handle more than one at
the same time in case we get another event before the previous
one finished

Clear offscreen on draw method (fix glitches)

Add child_resize JS method

Commit:
Rename project to maxwell-web-view

Commit:
JS: use "maxwell" string as main entry point

Add maxwell prefix to script messages

Commit:
Rename MaxwellWebViewChild struct as ChildData

Commit:
Rename package to maxwell

Commit:
Use 'GtkWidget' as class instead of 'MaxwellWebViewChild' for canvas elements

Check canvas-id is unique

Commit:
Add Readme file

Commit:
Add introspection support

Commit:
Add JS example

cosimoc commented on README.md:
Missing terminal period

cosimoc commented on README.md:
Missing terminal period

cosimoc commented on src/meson.build:
```diff
@@ -0,0 +1,88 @@
+maxwell_sources = [
+  'maxwell.c',
+  'maxwell-web-view.c',
+  'js-utils.c',
```
Don't you need to add `js-utils.h` to any of the sources array here?

ebassi replied:
You never need to add headers to source lists in Meson; the dependency is automatic.

This is technically true for Autotools as well, but Autotools needs the headers — even private ones — in order to do dist.

cosimoc commented on src/maxwell.c:
```diff
@@ -0,0 +1,8 @@
+#include "maxwell.h"
```
Add the header here as well

cosimoc replied:
Still needs addressing

cosimoc commented on src/maxwell-web-view.h:
```diff
@@ -0,0 +1,40 @@
+#define MAXWELL_TYPE_WEB_VIEW (maxwell_web_view_get_type ())
+G_DECLARE_FINAL_TYPE (MaxwellWebView, maxwell_web_view, MAXWELL, WEB_VIEW, WebKitWebView)
+
+GtkWidget     *maxwell_web_view_new ();
```
I'd add `void` as the parameter here.

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+ * maxwell_web_view_pack_child:
+ * @webview: a #MaxwellWebView
+ * @child: a #GtkWidget
+ * id: a string identifying child
```
Should be `@id`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+ * This is the same as calling gtk_container_add() and setting "canvas-id"
+ * packing property to @id.
+ *
+ * Since: 1.0
```
This library is not 1.0 yet, is it? I wonder if we can remove all the `Since` tags for now.

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+ * Since: 1.0
+ */
+GtkWidget *
+maxwell_web_view_new ()
```
Add `void`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+GtkWidget *
+maxwell_web_view_new ()
+{
+  return (GtkWidget *) g_object_new (MAXWELL_TYPE_WEB_VIEW, NULL);
```
`g_object_new()` does not need a cast here, since it returns a `gpointer`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+  GdkWindow *window = gtk_widget_get_window (widget);
+  if (gtk_cairo_should_draw_window (cr, window))
+    {
+      GTK_WIDGET_CLASS (maxwell_web_view_parent_class)->draw (widget, cr);
```
I would prefer an early return here.

xjuan replied:
Actually it should not be exclusive it should check every window

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+                                data->id, data->alloc.width, data->alloc.height);
+    }
+
+  _js_run_string (WEBKIT_WEB_VIEW (widget), script);
```
I'd find it clearer if the string was freed after this call.

xjuan replied:
Right I do not like this function either, but the JS is run asynchronously and I really like to print which JS script failed on error
Maybe I should rename the function to _js_run_take_string() ? to be clear that it takes ownership of it?

cosimoc replied:
Yeah, that would make it clearer.

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+maxwell_web_view_damage_event (GtkWidget *widget, GdkEventExpose *event)
+{
+  MaxwellWebViewPrivate *priv = MAXWELL_WEB_VIEW_PRIVATE (widget);
+  ChildData *data = get_child_data_by_offscreen (priv, event->window);
```
I would call this after checking `priv->script_loaded`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+      gtk_widget_unregister_window (widget, data->offscreen);
+      gdk_window_destroy (data->offscreen);
+      data->offscreen = NULL;
```
Use `g_clear_pointer()`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+      ensure_offscreen (widget, data);
+
+      if (gtk_widget_get_visible (data->child))
+        child_update_visibility (MAXWELL_WEB_VIEW (widget), l->data);
```
The second parameter to this should be `data->child`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+      ensure_offscreen (widget, data);
+
+      if (gtk_widget_get_visible (data->child))
```
Can't you call this unconditionally? The code should do the right thing even if child isn't visible.

xjuan replied:
Yes, I was just trying to reduce the calls to JS, but I plant to batch all the calls in one

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+  G_OBJECT_CLASS (maxwell_web_view_parent_class)->constructed (object);
+
+  /* Give XMlHttpRequest a chance to work */
```
Typo: `XmlHttpRequest`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+  /* Add script */
+  content_manager = webkit_web_view_get_user_content_manager (webview);
+  script_source = g_resources_lookup_data (RESOURCES_PATH"/maxwell-web-view.js",
```
This should be unreffed at the end

xjuan replied:
Good catch, btw I plan to move to a different place, that only gets called once when you actually try to add a widget, that way we dont inject extra JS on webview that do not plan to have children

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+  script_source = g_resources_lookup_data (RESOURCES_PATH"/maxwell-web-view.js",
+                                           G_RESOURCE_LOOKUP_FLAGS_NONE,
+                                           NULL);
+  script = webkit_user_script_new (g_bytes_get_data (script_source, NULL),
```
Pretty sure this should also be unreffed after being added to the user content manager.

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+  JSGlobalContextRef context = webkit_javascript_result_get_global_context (result);
+  JSValueRef value = webkit_javascript_result_get_value (result);
+
+  if (JSValueIsArray (context, value))
```
Return early if this fails

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+      priv->pixbufs = g_list_prepend (priv->pixbufs, pixbuf);
+
+      _js_run (WEBKIT_WEB_VIEW (widget),
+               "maxwell.child_draw('%s', '%s', %d, %d, %d, %d);",
```
Add space after `draw`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+          if (data)
+            {
+              child_allocate (data);
```
Do we not end up calling `gtk_widget_size_allocate()` outside of the allocation cycle as a consequence of this?

xjuan replied:
Yes, I will see if I can simplify this, adding a comment for now

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+  JSGlobalContextRef context = webkit_javascript_result_get_global_context (result);
+  JSValueRef value = webkit_javascript_result_get_value (result);
+
+  if (JSValueIsArray (context, value))
```
Return early if this fails

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+  webkit_uri_scheme_request_finish_error (request,
+                                          g_error_new_literal (MAXWELL_ERROR, 0,
+                                                               "Could not find imagedata uri"));
```
I believe that the error may get leaked in this case.

xjuan replied:
ouch, it does

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+       * Where 'id' is the child id and 'pixbuf' is an optional GdkPixbuf
+       * pointer encoded in base64
+       *
+       * GdkPixbuf are created in damage-event handler and pusshed to priv->pixbufs
```
Typo: `pusshed`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+                             pixbuf);
+
+          /* Pixbuf is no longer our responsibility, stream will take care
+           * of freeing it when its finalized
```
Typo: `it's finalized`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,898 @@
+
+static GParamSpec *child_properties[N_CHILD_PROPERTIES];
+
+G_DEFINE_TYPE_WITH_PRIVATE (MaxwellWebView, maxwell_web_view, WEBKIT_TYPE_WEB_VIEW);
```
No need for the terminal semicolon

cosimoc commented on meson_options.txt:
```diff
@@ -0,0 +1,3 @@
+option('introspection', type: 'boolean',
```
Why would one ever want to disable introspection for this library?

xjuan replied:
I do not know, a C program that wants to link with its own copy of the library but does not want to depend on gobject introspection to build it

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+/* Semi Public API */
+
+/* Main entry point */
+window.maxwell = {}
```
Missing semicolon

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+window.maxwell = {}
+
+/* List of children */
+window.maxwell.children = {}
```
Missing semicolon

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+    window.maxwell.child_draw (id, null, 0, 0, width, height);
+
+    if (visible)
+        window.maxwell.child_set_visible (id, visible);
```
Use a consistent JS coding style for whether or not a space exists before braces when calling a method

Commit:
Always build introspection, remove option.

Commit:
Fix JS style and add missing semicolons

Commit:
Add logo to readme file

Commit:
MaxwellWebView: various fixes and code cleanups

on_image_data_uri_scheme_request() fix GError leak

maxwell_web_view_constructed() fix WebKitUserScript and script_source leak

maxwell_web_view_draw() always check every window if we should draw it

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+        return;
+
+    /* Resize canvas keeping the old contents */
+    if (child.width !== width || child.height !== height) {
```
Nitpick: prefer an early return here

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+        child.width = width;
+        child.height = height;
+
+        ctx.putImageData(old_image, 0, 0);
```
I'm not sure I fully understand the logic here; you're fetching a `child.width X child.height` snapshot of the children, but then paint the same image data in the same exact position.
If `child` got larger, that should have no effect. If `child` got smaller, would it not be much faster to simply clip to the new size? What am I missing?

xjuan replied:
That code used to be part of draw_child() and I was trying to avoid flickering (broadway does the same)
Anyways the canvas elements get cleared when its resized but yeah I could simply get the new image instead

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+            child.maxwell_display_value = child.style.display;
+
+            /* Hide all widgets by default */
+            child.style.display = 'none';
```
Would be nice to perhaps re-use the visibility helper function you created for this.

xjuan replied:
rigth, but this is only initializing the visibility state.
The idea is to hide everything from the DOM tree until gtk can take control of it.

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,242 @@
+
+    /* Update all positions in MaxwellWebView at once to reduce messages */
+    window.webkit.messageHandlers.maxwell_update_positions.postMessage(positions);
+    position_timeout_id = null;
```
This does not seem to be defined anywhere

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,897 @@
+        }
+    }
+
+  error = g_error_new_literal (MAXWELL_ERROR, 0,
```
Instead of `0`, I'd like an enum of maxwell error codes.  It's fine if for now you only have a generic "failed" error, but the enum value should be used here.

Commit:
Add MaxwellError enumeration

Use MAXWELL_ERROR_URI for custom uri handler errors

Commit:
Improve JS run convenience functions

Rename _js_run() as _js_run_printf()

Add convenince macros that include the function name for asynch error msgs

Commit:
JS: clean up child_resize() and child_draw()

Add get_image() function to be able to get the image before resizing the canvas

xjuan commented:
Ok so I think the only thing left is to cleanup the child allocation process, but I think that deserve it's own task

cosimoc commented on src/js-utils.c:
```diff
@@ -0,0 +1,117 @@
+  webkit_web_view_run_javascript (WEBKIT_WEB_VIEW (webview),
+                                  script, NULL,
+                                  _js_run_finish_handler,
+                                  (gpointer)function);
```
You're not guaranteed that the memory pointed to by `function` will still be valid by the time the async function has returned, so you should probably copy the string before passing it (and then freeing it in the callback)

xjuan replied:
Ohh right, __func__ is not a literal!

cosimoc commented on src/js-utils.c:
```diff
@@ -0,0 +1,117 @@
+  gchar *function = data;
+  GError *error = NULL;
+
+  js_result = webkit_web_view_run_javascript_finish (WEBKIT_WEB_VIEW (object),
```
This is leaked

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,903 @@
+                                  data->offscreen);
+
+  if (data->offscreen)
+    gdk_window_destroy (data->offscreen);
```
You could use `g_clear_pointer()`

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,903 @@
+  MaxwellWebViewPrivate *priv = MAXWELL_WEB_VIEW_PRIVATE (self);
+
+  priv->children = NULL;
+  priv->pixbufs  = NULL;
```
These should not be needed, the priv structured is initialized with zeroes anyway.

cosimoc commented on src/maxwell-web-view.c:
```diff
@@ -0,0 +1,903 @@
+  MaxwellWebViewPrivate *priv = MAXWELL_WEB_VIEW_PRIVATE (widget);
+  ChildData *data;
+
+  /* Dont do anything if the support script did not finished loading */
```
Nit: `Don't`

cosimoc commented on src/maxwell-web-view.js:
```diff
@@ -0,0 +1,239 @@
+    let child = maxwell.children[id];
+
+    if (!child)
+        return;
```
All the functions that this calls already check that `maxwell.children[id]` is not undefined. Since `child` here is not used directly, perhaps we can remove this code.

cosimoc commented:
@xjuan Only few minor comments left. I agree to split the allocation stuff for a follow-up ticket/PR.

Commit:
Fix leaks in _js_run_finish_handler() and others minor cleanups

cosimoc reviewed (approved):
Thanks, @xjuan. This looks good to me now. Not sure if there's anything left to backport here into master, but feel free to do so and close this PR if so.

Commit:
Remove private headers from build system.

xjuan commented:
Ok good, closing this PR
Pending improvements moved to https://phabricator.endlessm.com/T21883
